src/outpack/location_pull.py

```diff
@@ -0,0 +1,463 @@
+            mark_known(
+                root,
+                packet_id,
+                location_name,
+                known_there[packet_id].hash,
+                known_there[packet_id].time,
+            )
+
+
+def outpack_location_pull_packet(
+    ids: Union[str, List[str]],
+    options=None,
+    recursive=None,
+    root=None,
+    *,
+    locate=True,
+):
+    root = root_open(root, locate=locate)
+    options = SearchOptions(options)
+    if isinstance(ids, str):
+        ids = [ids]
+    plan = _location_build_pull_plan(ids, options.location, recursive, root)
+
+    ## Warn people of extra pulls and skips
+    if plan.info.n_extra > 0:
+        print(
+            f"Also pulling {plan.info.n_extra} "
+            f"{pl(plan.info.n_extra, 'packet')}, which are "
+            f"dependencies of those requested"
+        )
+    if plan.info.n_skip > 0:
+        print(
+            f"Skipping {plan.info.n_skip} of {plan.info.n_total} "
+            f"{pl(plan.info.n_total, 'packet')} which are already "
+            f"unpacked"
+        )
+
+    store, cleanup = _location_pull_files(plan.files, root)
+
+    use_archive = root.config.core.path_archive is not None
+    n_packets = len(plan.packets)
+    time_start = time.time()
+    for idx, packet in enumerate(plan.packets.values()):
+        if use_archive:
+            print(
+                f"Writing files for '{packet.packet}' (packet {idx + 1}/"
+                f"{n_packets})"
+            )
+            _location_pull_files_archive(packet.packet, store, root)
+
+        mark_known(
+            root, packet.packet, LOCATION_LOCAL, packet.hash, time.time()
+        )
+
+    print(
+        f"Unpacked {n_packets} {pl(n_packets, 'packet')} in "
+        f"{humanize.time.precisedelta(int(time.time() - time_start))}."
+    )
+
+    cleanup()
+    return list(plan.packets.keys())
+
+
+# This approach may be suboptimal in the case where the user does not
+# already have a file store, as it means that files will be copied
+# around and hashed more than ideal:
+#
+# * hash the candidate file
+# * rehash on entry into the file store
+# * copy into the file store
+# * copy from the file store into the final location
+#
+# So in the case where a hash is only present once in a chain of
+# packets being pulled this will be one too many hashes and one too
+# many copies.
+#
+# However, this approach makes the logic fairly easy to deal with,
+# and copes well with data races and corruption of data on disk
+# (e.g., users having edited files that we rely on, or editing them
+# after we hash them the first time).
+def _location_pull_files(
+    files: List[PacketFileWithLocation], root: OutpackRoot
+) -> Tuple[FileStore, Callable[[], None]]:
+    store = root.files
+    if store is not None:
+
+        def cleanup():
+            return None
+
+        exists, missing = partition(lambda file: store.exists(file.hash), files)
+
+        if len(exists) > 0:
```

Comment:
```suggestion
        if exists:
```

I think this is the pythonic way of doing these tests